cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java

```diff
@@ -1516,25 +1529,25 @@ private static void templates(LexerfulGrammarBuilder b) {
 
     b.rule(templateArgument).is(
       b.firstOf(
-        typeId,
-        
+        b.sequence(typeId, b.next(b.firstOf(">", ",", "..."))),
+        b.sequence(typenameSpecifier, b.next(b.firstOf(">", ",", "..."))), // seen in gnu system headers
         // FIXME: workaround to parse stuff like "carray<int, 10>"
         // actually, it should be covered by the next rule (constantExpression)
         // but it doesnt work because of ambiguity template syntax <--> relationalExpression
-        shiftExpression,
-        constantExpression,
-        idExpression
+        b.sequence(shiftExpression, b.next(b.firstOf(">", ","))),
+        b.sequence(constantExpression, b.next(b.firstOf(">", ",")))
+//        idExpression
       )
```

Comment:
I have left out id-expressions for the rule for template-argument because it is already a special case of a constant-expression within the grammar implementation of sonar-cxx. 
With the chosen ordering the parser performs remarkably well in spite of not keeping a symbol table and not doing backtracking.